src/front/glsl_new/lex.rs

```diff
@@ -0,0 +1,314 @@
+                //     chars.as_str(),
+                //     start,
+                //     start + 2,
+                // ),
+                _ => (Some(Token::MulAssign(meta)), input),
+            }
+        }
+        '/' => {
+            input = chars.as_str();
+            match chars.next() {
+                Some('=') => {
+                    meta.chars.end = start + 2;
+                    (Some(Token::DivAssign(meta)), chars.as_str())
+                }
+                //TODO: line comments
+                // Some('/') => (Token::LineComment, chars.as_str(), start, start + 2),
+                //TODO: multi-line comments
+                // Some('*') => (
+                //     Token::MultiLineCommentOpen,
+                //     chars.as_str(),
+                //     start,
+                //     start + 2,
+                // ),
+                _ => (Some(Token::Slash(meta)), input),
+            }
+        }
+        '#' => {
+            input = chars.as_str();
+            let (word, rest, pos) = consume_any(input, |c| c.is_alphanumeric() || c == '_');
+            meta.chars.end = start + pos;
+            match word {
+                "version" => (Some(Token::Version(meta)), rest),
+                _ => (None, input),
+            }
+
+            //TODO: preprocessor
+            // if chars.next() == Some(cur) {
+            //     (Token::TokenPasting, chars.as_str(), start, start + 2)
+            // } else {
+            //     (Token::Preprocessor, input, start, start + 1)
+            // }
+        }
+        '~' => (Some(Token::Tilde(meta)), chars.as_str()),
+        '?' => (Some(Token::Question(meta)), chars.as_str()),
+        _ => (None, chars.as_str()),
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Lexer<'a> {
+    lines: Enumerate<Lines<'a>>,
+    input: String,
+    line: usize,
+    offset: usize,
+}
+
+impl<'a> Lexer<'a> {
+    pub fn new(input: &'a str) -> Self {
+        let mut lines = input.lines().enumerate();
+        let (line, input) = lines.next().unwrap_or((0, ""));
+        let mut input = String::from(input);
+
+        while input.ends_with('\\') {
+            if let Some((_, next)) = lines.next() {
+                input.pop();
+                input.push_str(next);
+            } else {
+                break;
+            }
+        }
+
+        Lexer {
+            lines,
+            input,
+            line,
+            offset: 0,
+        }
+    }
+
+    #[must_use]
+    pub fn next(&mut self) -> Option<Token> {
+        let (token, rest) = consume_token(&self.input);
+
+        if let Some(mut token) = token {
+            self.input = String::from(rest);
+            let meta = token.extra_mut();
+            let end = meta.chars.end;
+            meta.line = self.line;
+            meta.chars.start += self.offset;
+            meta.chars.end += self.offset;
+            self.offset += end;
+            Some(token)
+        } else {
+            match self.lines.next() {
```

Comment:
nit: can do `let (line, input) = self.lines.next()?;` and proceed without a match

Reply:
Done